### PR TITLE
Issue 6311 : Admin CLI Commands to view Reader Information in zookeeper for a specific Controller host

### LINKED ISSUE #6311: Troubleshooting commands for inspecting the state of a Stream, Reader Group, Key Value Table or Transaction
**Is your feature request related to a problem? Please describe.**
We need a simple way to inspect the state of a specific Stream, Reader Group, Key Value Table or Transaction in a Pravega system at any point during its operation.

**Describe the solution you'd like**
Troubleshooting commands for inspecting the state of a Stream, Reader Group, Key Value Table or Transaction

**Describe alternatives you've considered**
TBD

**Additional context**
TBD

### PR DESCRIPTION
**Pull request by Bhupender-Y**
Signed-off-by: Bhupender-Y <Bhupender.Y@dell.com>

**Change log description**  
Add a command `controller-metadata getReader <host-id> <reader-group> <reader-id>` in admin-cli to view metadata of reader of a particular controller instance.

**Purpose of the change**  
Fixes #6311

**What the code does**  
 Created  `controller-metadata getReader <host-id> <reader-group> <reader-id>`  command in admin cli to view reader metadata.

**How to verify it**  
- Navigate to pravega-admin cli.
- Use `controller-metadata getReader <host-id> <reader-group> <reader-id>`. It should print reader metadata.

### CODE REVIEW

File: cli/admin/src/main/java/io/pravega/cli/admin/utils/ZKHelper.java [hunk 1]
```diff
@@ -112,6 +117,22 @@ private ZKHelper(String zkURL, String clusterName) throws ZKConnectionFailedExce
 
     }
 
+    /**
+     * Get the metadata information of reader for a given path.
+     * @param readerPath path of reader
+     * @throws Exception if unable to get the znode for readerpath.
+     * @return detailed metadata information of reader
+     */
+    public String getMetaDataForReader(String readerPath) throws Exception {
```
Comment (shshashwat) [inline cli/admin/src/main/java/io/pravega/cli/admin/utils/ZKHelper.java:126]: Is it possible to be specific about the Exception thrown here and how is it handled?

**Review by shshashwat — CHANGES_REQUESTED** [review]

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataCommand.java [hunk 2]
```diff
@@ -187,4 +188,17 @@ ByteBuffer getByteBuffer(ByteBuf byteBuf) {
         byteBuf.getBytes(readerIndex, bytes);
         return ByteBuffer.wrap(bytes);
     }
+
+
+    /**
+     * Method to get path of reader in a particular controller instance.
+     *
+     * @param process Host id of controller instance
```
[the remark this answers is not in the document]
Comment (Bhupender-Y) [reply cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataCommand.java]: Changed as per suggestion.

File: cli/admin/src/test/java/io/pravega/cli/admin/controller/AbstractControllerMetadataCommandsTest.java [hunk 3]
```diff
@@ -307,6 +312,33 @@ public void testUpdateControllerMetadataEntryCommandJSONSyntaxException() throws
...
     }
 
+    @Test
+    public void testControllerMetadataViewReaderInfoCommand() throws Exception {
+        final String process = UUID.randomUUID().toString();
+        final String readerGroup = UUID.randomUUID().toString();
+        final String reader = UUID.randomUUID().toString();
+        ZKHelper zkHelper = ZKHelper.create(SETUP_UTILS.getZkTestServer().getConnectString(), "pravega-cluster");
+        CheckpointStore checkpointStore = zkHelper.getCheckPointStore();
+        checkpointStore.addReaderGroup(process, readerGroup);
```
[the remark this answers is not in the document]
Comment (Bhupender-Y) [reply cli/admin/src/test/java/io/pravega/cli/admin/controller/AbstractControllerMetadataCommandsTest.java:322]: getCheckpointStore will never give an exception as it is handled in CheckpointStoreFactory. So removed extra try-catch block.

File: cli/admin/src/main/java/io/pravega/cli/admin/utils/ZKHelper.java [hunk 4]
```diff
@@ -155,6 +176,20 @@ public static ZKHelper create(String zkURL, String clusterName) throws ZKConnect
...
+    /**
+     * Get the checkPointStore.
+     * @return checkPointStore
+     */
+    public CheckpointStore getCheckPointStore() {
+        CheckpointStore checkpointStore = null;
+        try {
+            checkpointStore = CheckpointStoreFactory.createZKStore(zkClient);
+        } catch (Exception e) {
+            System.err.println("An error occurred executing getCheckPointStore against Zookeeper: " + e.getMessage());
```
[the remark this answers is not in the document]
Comment (Bhupender-Y) [reply cli/admin/src/main/java/io/pravega/cli/admin/utils/ZKHelper.java]: getCheckpointStore will never give an exception as it is handled in CheckpointStoreFactory. So removed extra try-catch block.

File: cli/admin/src/main/java/io/pravega/cli/admin/utils/ZKHelper.java (hunk 1 above)
Comment (Bhupender-Y) [reply cli/admin/src/main/java/io/pravega/cli/admin/utils/ZKHelper.java:126]: forPath() method throws Exception thats why proceed it further. It is handled [here](https://github.com/Bhupender-Y/pravega/blob/issue-6311-admin-cli-command-to-view-reader-info/cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataViewReaderInfoCommand.java#L41)

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataViewReaderInfoCommand.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        super(args);
+    }
+
+    @Override
+    public void execute() throws Exception {
+        ensureArgCount(3);
+
+        final String hostId = getArg(0);
+        final String readerGroupName = getArg(1);
+        final String readerId = getArg(2);
```
Comment (abhinb) [inline cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataViewReaderInfoCommand.java:39]: is there anything to be validated once we take these inputs?

File: cli/admin/src/main/java/io/pravega/cli/admin/utils/ZKHelper.java (hunk 1 above)
Comment (abhinb) [reply cli/admin/src/main/java/io/pravega/cli/admin/utils/ZKHelper.java:126]: what gets show in cli in case of exception?

Comment (shshashwat) [reply cli/admin/src/main/java/io/pravega/cli/admin/utils/ZKHelper.java:126]: can you check the behavior in case of exception via the test case?

Comment (Bhupender-Y) [reply cli/admin/src/main/java/io/pravega/cli/admin/utils/ZKHelper.java:126]: When an exception occurs, **Exception accessing to reader metadata** message shows in cli. It is handled [here](https://github.com/Bhupender-Y/pravega/blob/issue-6311-admin-cli-command-to-view-reader-info/cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataViewReaderInfoCommand.java#L41).  To verify exception, **testControllerMetadataViewReaderInfoCommandWithException** is written in [AbstractControllerMetadataCommandsTest ](https://github.com/Bhupender-Y/pravega/blob/issue-6311-admin-cli-command-to-view-reader-info/cli/admin/src/test/java/io/pravega/cli/admin/controller/AbstractControllerMetadataCommandsTest.java#L331).

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataViewReaderInfoCommand.java (hunk 5 above)
Comment (Bhupender-Y) [reply cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataViewReaderInfoCommand.java:39]: No need to validate the inputs. If inputs is wrong, then will not find zNode at given path and exception message will shows at cli.

**Comment by anirudhkovuru** [comment]
@Bhupender-Y do update the admin CLI's documentation (the README) with the new command.

**Review by pbelgundi — CHANGES_REQUESTED** [review]
@Bhupender-Y Could you please paste the output of manual testing for this command in the PR?

**Review by anirudhkovuru — APPROVED** [review]
LGTM

**Comment by Bhupender-Y** [comment]
Output of command will look like 
**reader-metadata : 
 _system/_requeststream/0.#epoch.0=187, _system/_requeststream/1.#epoch.0=0**

**Review by pbelgundi — APPROVED** [review]
LGTM

**Review by shshashwat — APPROVED** [review]

**Pull request merged**